### ISSUE TITLE
Pass CPython's "test_pickle"

### ISSUE THREAD

**Issue by kmod** [pyston/pyston#918]
We don't have it loaded into our repo, but it would be good to look into how we do on this.  (I just noticed that we don't correctly pickle subclasses of tuple, and presumably subclasses of other builtins.)

**Comment by kmod** [comment]
I have a pr forthcoming that improves our pickling support (and adds test_pickle.py) but it depends on #920.

**Comment by Daetalus** [comment]
With some latest PRs, the "test_pickle" almost could work now. Except one test failure:

```
import pickle
mymodule = pickle
builtins = {mymodule.__name__: mymodule,
            '__import__': __import__}

teststr = '''def f(): {0}.dumps(0)'''.format(mymodule.__name__)
exec teststr in {'__builtins__': builtins}, d
d['f']()
```
Seems the module scope has a problem. The `pickle.dumps(0)` inside `teststr` could not recognize the `pickle` correctly. PyPy also has same issue.

BTW: During improve Pyston compatibility, I found that PyPy also has a lot of compatibility issues. But rarely seen relevant reports. Maybe just few people use it in production?

**Comment by kmod** [comment]
Oh erf, yeah we don't support requesting a custom `__builtins__` dict.  I just opened #946 which at least adds an assert if we run into this case, so it'll be more clear what's going on.

Let's hope that we can get away with not supporting this :)  CPython's rules around `__builtins__` seem kind of ad-hoc so it would be tough for people to rely on them, and the fact that PyPy doesn't support this is evidence that people don't care about it.  Maybe we can just skip that test?

I've filed bugs against PyPy that originate from our test suite :)  I found [this](https://bitbucket.org/pypy/pypy/issues/876/__builtins__-aint-modifyable-in-pypy-but) which seems somewhat related to this issue you ran into, but yeah it definitely feels like we're reaching their level of compatibility (if something doesn't work for us it has a good chance of not working for them).  I don't know how many people use PyPy in production, but I think it's small enough that there is a lot we can do for Python performance :)
